presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseCostBasedPlanTest.java

```diff
@@ -0,0 +1,240 @@
+import static java.nio.file.Files.exists;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static org.testng.Assert.assertEquals;
+
+public abstract class BaseCostBasedPlanTest
+        extends BasePlanTest
+{
+    public BaseCostBasedPlanTest(LocalQueryRunnerSupplier supplier)
+    {
+        super(supplier);
+    }
+
+    protected abstract Stream<Query> getQueries();
+
+    protected abstract Path getExpectedJoinOrderingFile(String queryId);
+
+    @DataProvider
+    public Object[][] getQueriesDataProvider()
+    {
+        return getQueries()
+                .collect(toDataProvider());
+    }
+
+    @Test(dataProvider = "getQueriesDataProvider")
+    public void test(Query query)
+    {
+        assertEquals(generateQueryPlan(resolve(query.queryFile)), read(resolve(getExpectedJoinOrderingFile(query.id))));
+    }
+
+    private Path resolve(Path path)
+    {
+        // Test is run with module as current working directory. generate() is run with project's top-level directory as current working directory.
+        if (exists(path)) {
+            return path;
+        }
+        Path fromParent = Paths.get("..").resolve(path);
+        checkState(exists(fromParent), "Unable to resolve: " + path);
+        return fromParent;
+    }
+
+    public void generate()
+            throws Exception
+    {
+        initPlanTest();
+        try {
+            getQueries()
+                    .parallel()
+                    .forEach(query -> {
+                        try {
+                            Files.write(generateQueryPlan(query.queryFile).getBytes(UTF_8), getExpectedJoinOrderingFile(query.id).toFile());
+                            System.out.println(query.id);
+                        }
+                        catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                    });
+        }
+        finally {
+            destroyPlanTest();
+        }
+    }
+
+    private static String read(Path file)
+    {
+        try {
+            return Files.asCharSource(file.toFile(), UTF_8).read();
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static final class Query
+    {
+        private final String id;
+        private final Path queryFile;
+
+        public Query(String id, Path queryFile)
+        {
+            this.id = requireNonNull(id, "id is null");
+            this.queryFile = requireNonNull(queryFile, "queryFile is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            // Keep toString() short, as this is used in @DataProvider
+            return id;
+        }
+    }
+
+    private String generateQueryPlan(Path queryFile)
+    {
+        String sql = read(queryFile)
+                .replaceFirst(";", "")
```

Comment:
Please just copy the queries here. After you have the queries in, introduce some convention, and remove all the weird methods for resolving the queries/results. Once you have them in a single folder, you can scan the classpath, run everything under this directory and compare it with the results stored in some other directory under the classpath.

Reply:
It is about 125 queries which we modify from time to time. The goal was to test exactly what we benchmark. Once we copy them here it will be no longer guaranteed.